include/libp2p/protocol/gossip/impl/message_cache.hpp

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LIBP2P_PROTOCOL_GOSSIP_MESSAGE_CACHE_HPP
+#define LIBP2P_PROTOCOL_GOSSIP_MESSAGE_CACHE_HPP
+
+#include <functional>
+
+#include <boost/multi_index/hashed_index_fwd.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index_fwd.hpp>
+#include <boost/multi_index_container_fwd.hpp>
+
+#include <libp2p/protocol/gossip/common.hpp>
+
+namespace libp2p::protocol::gossip {
+
+  struct ById;
+  struct ByTimestamp;
+
+  namespace msg_cache_table {
+
+    namespace mi = boost::multi_index;
+
+    struct Record {
+      MessageId message_id;
+      Time inserted_at;
+      TopicMessage::Ptr message;
+    };
+
+    /// Table of Record with 2 indices (by key and expire time)
+    using Table = boost::multi_index_container<
+        Record,
+        mi::indexed_by<
+            // hashed_non_unique by id means unordered index
+            mi::hashed_unique<
+                mi::tag<ById>,
+                mi::member<Record, MessageId, &Record::message_id>,
+                std::hash<MessageId>>,
+            // ordered_non_unique by time means accending order
+            mi::ordered_non_unique<
+                mi::tag<ByTimestamp>,
+                mi::member<Record, Time, &Record::inserted_at>>>>;
+
+  }  // namespace msg_cache_table
+
+  /// Message cache with expiration
+  class MessageCache {
+   public:
+    using TimeFunction = std::function<Time()>;
+
+    MessageCache(Time message_lifetime, Time broadcast_lifetime,
+                 TimeFunction clock);
+
+    ~MessageCache();
+
+    /// Callback for retrieving data for IHave protocol notifications
+    using IHaveCallback =
+    std::function<void(const TopicId &, const MessageId &)>;
+
+    /// Forwards actual seen message ids to callback by topic given
+    /// or all pairs if topic==""
+    void getSeenMessageIds(const TopicId &topic,
+                           const IHaveCallback &callback) const;
+
+    /// Returns message by id if found, or empty ptr otherwise
+    TopicMessage::Ptr getMessage(const MessageId &id) const;
```

Comment:
It is not obvious for users of method that ptr can be empty. Replace with optional on ptr, so that ptr is never nullptr
```suggestion
    boost::optional<TopicMessage::Ptr> getMessage(const MessageId &id) const;
```